CodeGeneration/Sources/generate-swiftsyntax/templates/swiftsyntax/SyntaxNodeFile.swift

```diff
@@ -0,0 +1,494 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftSyntax
+import SwiftSyntaxBuilder
+import SyntaxSupport
+import Utils
+
+/// This gyb-file generates the syntax nodes for SwiftSyntax. To keep the generated
+/// files at a managable file size, it is to be invoked multiple times with the
+/// variable `emitKind` set to a base kind listed in
+/// It then only emits those syntax nodes whose base kind are that specified kind.
+func syntaxNode(emitKind: String) -> SourceFileSyntax {
+  SourceFileSyntax(leadingTrivia: .docLineComment(generateCopyrightHeader(for: "generate-swiftsyntax"))) {
+    for node in SYNTAX_NODES where !node.isBase && node.collectionElement.isEmpty && node.baseKind == emitKind {
+      // We are actually handling this node now
+      let nodeDoc = node.description.map { "/// \($0)" }
+      try! StructDeclSyntax(
+        """
+        // MARK: - \(raw: node.name)
+
+        \(raw: nodeDoc ?? "")
+        public struct \(raw: node.name): \(raw: node.baseType.syntaxBaseName)Protocol, SyntaxHashable
+        """
+      ) {
+        for child in node.children {
+          if let childChoiceDecl = try generateSyntaxChildChoices(for: child) {
+            childChoiceDecl
+          }
+        }
+
+        // ==============
+        // Initialization
+        // ==============
+
+        DeclSyntax("public let _syntaxNode: Syntax")
+
+        DeclSyntax(
+          """
+          public init?<S: SyntaxProtocol>(_ node: S) {
+            guard node.raw.kind == .\(raw: node.swiftSyntaxKind) else { return nil }
+            self._syntaxNode = node._syntaxNode
+          }
+          """
+        )
+
+        DeclSyntax(
+          """
+          /// Creates a `\(raw: node.name)` node from the given `SyntaxData`. This assumes
+          /// that the `SyntaxData` is of the correct kind. If it is not, the behaviour
+          /// is undefined.
+          internal init(_ data: SyntaxData) {
+            assert(data.raw.kind == .\(raw: node.swiftSyntaxKind))
+            self._syntaxNode = Syntax(data)
+          }
+          """
+        )
+
+        try! InitializerDeclSyntax("\(node.generateInitializerDeclHeader(optionalBaseAsMissing: false))") {
+          let parameters = node.children.map { $0.swiftName }.joined(separator: ", ")
```

Comment:
I know that I wrote this but reading it again, we should be using syntax nodes here as well:

```swift
let parameters = FunctionParameterListSyntax {
  for child in node.children {
    FunctionParameterSyntax(firstName: child.swiftName)
  }
}
```